src/terminal-util.c

```diff
@@ -707,6 +708,55 @@ terminal_util_add_proxy_env (GHashTable *env_table)
 	g_object_unref (settings_socks);
 }
 
+/**
+ * terminal_util_hyperlink_uri_label:
+ * @uri: a URI
+ *
+ * Formats @uri to be displayed in a tooltip.
+ * Performs URI-decoding and converts IDN hostname to UTF-8.
+ *
+ * Returns: (transfer full): The human readable URI as plain text
+ */
+char *terminal_util_hyperlink_uri_label (const char *uri)
+{
+	char *unesc = NULL;
+	gboolean replace_hostname;
+
+	if (uri == NULL)
+		return NULL;
+
+	unesc = g_uri_unescape_string(uri, NULL);
+	if (unesc == NULL)
+		unesc = g_strdup(uri);
+
+	if (g_ascii_strncasecmp(unesc, "ftp://", 6) == 0 ||
+		g_ascii_strncasecmp(unesc, "http://", 7) == 0 ||
+		g_ascii_strncasecmp(unesc, "https://", 8) == 0)
```

Comment:
Wondering whether `mailto:` addresses can be IDN. Probably an overkill to add more.